### PR TITLE
[deckhouse-controller] Fixed update.windows for the new DeckhouseRelease update logic

### PR DESCRIPTION
**Pull request by libmonsoon-dev**
## Description
Fixed update logic in various modes

## Why do we need it, and what problem does it solve?
Close https://github.com/deckhouse/deckhouse/issues/10000

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: deckhouse-controller
type: fix
summary: Fixed update logic in various modes
impact_level: default
```

### CODE REVIEW

File: deckhouse-controller/pkg/controller/module-controllers/release/controller.go [hunk 1]
```diff
@@ -150,87 +149,82 @@ func NewModuleReleaseController(
...
 		}
 
-		c.m.Unlock()
+		r.m.Unlock()
 	}
 }
 
 // only ModuleRelease with active finalizer can get here, we have to remove the module on filesystem and remove the finalizer
-func (c *moduleReleaseReconciler) deleteReconcile(ctx context.Context, mr *v1alpha1.ModuleRelease) (ctrl.Result, error) {
-	res := ctrl.Result{}
```
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/module-controllers/release/controller.go:185]: please, return alias.
it makes code more readeble

**Review by ldmonster — CHANGES_REQUESTED** [review]

File: deckhouse-controller/pkg/controller/module-controllers/release/controller.go [hunk 2]
```diff
@@ -239,49 +233,49 @@ func (c *moduleReleaseReconciler) createOrUpdateReconcile(ctx context.Context, m
...
 	var metaUpdateRequired bool
 	// check if RegistrySpecChangedAnnotation annotation is set and processes it
 	if _, set := mr.GetAnnotations()[RegistrySpecChangedAnnotation]; set {
 		// if module is enabled - push runModule task in the main queue
-		c.logger.Infof("Applying new registry settings to the %s module", mr.Spec.ModuleName)
-		err := c.moduleManager.RunModuleWithNewOpenAPISchema(mr.Spec.ModuleName, mr.ObjectMeta.Labels["source"], filepath.Join(c.downloadedModulesDir, mr.Spec.ModuleName, fmt.Sprintf("v%s", mr.Spec.Version)))
+		r.logger.Infof("Applying new registry settings to the %s module", mr.Spec.ModuleName)
+		err := r.moduleManager.RunModuleWithNewOpenAPISchema(mr.Spec.ModuleName, mr.ObjectMeta.Labels["source"], filepath.Join(r.downloadedModulesDir, mr.Spec.ModuleName, fmt.Sprintf("v%s", mr.Spec.Version)))
 		if err != nil {
 			return ctrl.Result{Requeue: true}, err
```
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/module-controllers/release/controller.go:280]: if err != nil - result will be ignored

File: deckhouse-controller/pkg/controller/module-controllers/release/controller.go [hunk 3]
```diff
@@ -302,27 +296,27 @@ func (c *moduleReleaseReconciler) reconcileDeployedRelease(ctx context.Context,
 	}
 
 	if metaUpdateRequired {
-		return ctrl.Result{Requeue: true}, c.client.Update(ctx, mr)
```
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/module-controllers/release/controller.go:305]: we don't want to check update error and describe it?

File: deckhouse-controller/pkg/controller/module-controllers/release/controller.go [hunk 4]
```diff
@@ -399,41 +393,41 @@ func (c *moduleReleaseReconciler) reconcilePendingRelease(ctx context.Context, m
...
+			err = r.client.Get(ctx, types.NamespacedName{Name: policyName}, policy)
 			if err != nil {
-				c.metricStorage.CounterAdd("{PREFIX}module_update_policy_not_found", 1.0, map[string]string{
+				r.metricStorage.CounterAdd("{PREFIX}module_update_policy_not_found", 1.0, map[string]string{
 					"version":        mr.GetReleaseVersion(),
 					"module_release": mr.GetName(),
 					"module":         mr.GetModuleName(),
 				})
 
-				if e := c.updateModuleReleaseStatusMessage(ctx, mr, fmt.Sprintf("Update policy %s not found", policyName)); e != nil {
```
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/module-controllers/release/controller.go:414]: strange error name

File: deckhouse-controller/pkg/controller/module-controllers/release/controller.go [hunk 5]
```diff
@@ -399,41 +393,41 @@ func (c *moduleReleaseReconciler) reconcilePendingRelease(ctx context.Context, m
...
+				if e := r.updateModuleReleaseStatusMessage(ctx, mr, fmt.Sprintf("Update policy %s not found", policyName)); e != nil {
 					return ctrl.Result{Requeue: true}, e
 				}
 				return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
 			}
 		}
 
 		if policy.Spec.Update.Mode == "Ignore" {
-			if e := c.updateModuleReleaseStatusMessage(ctx, mr, disabledByIgnorePolicy); e != nil {
```
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/module-controllers/release/controller.go:422]: strange error name too

File: deckhouse-controller/pkg/controller/module-controllers/release/controller.go [hunk 6]
```diff
@@ -513,41 +512,43 @@ func (c *moduleReleaseReconciler) reconcilePendingRelease(ctx context.Context, m
...
+	var notReadyErr *updater.NotReadyForDeployError
+	if errors.As(err, &notReadyErr) {
+		r.logger.Infoln(err.Error())
+		// TODO: requeue all releases if deckhouse update settings is changed
+		// requeueAfter := notReadyErr.RetryDelay()
+		// if requeueAfter == 0 {
+		// requeueAfter = defaultCheckInterval
+		// }
+		// r.logger.Infof("%s: retry after %s", err.Error(), requeueAfter)
+		return ctrl.Result{Requeue: true, RequeueAfter: notReadyErr.RetryDelay()}, nil
```
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/module-controllers/release/controller.go:543]: ctrl.Result{RequeueAfter: notReadyErr.RetryDelay()}

if you fill RequeueAfter - its override Requeue: true.
Or we have possibility to RequeueAfter be zero?

Comment (libmonsoon-dev) [reply deckhouse-controller/pkg/controller/module-controllers/release/controller.go:545]: RetryDelay can be zero, but now all this logic is paused, this is a reserve for the future. Changed to defaultCheckInterval

File: deckhouse-controller/pkg/controller/module-controllers/release/controller.go (hunk 5 above)
Comment (libmonsoon-dev) [reply deckhouse-controller/pkg/controller/module-controllers/release/controller.go:422]: Fixed

File: deckhouse-controller/pkg/controller/module-controllers/release/controller.go (hunk 4 above)
Comment (libmonsoon-dev) [reply deckhouse-controller/pkg/controller/module-controllers/release/controller.go:414]: Fixed

File: deckhouse-controller/pkg/controller/module-controllers/release/controller.go (hunk 2 above)
Comment (libmonsoon-dev) [reply deckhouse-controller/pkg/controller/module-controllers/release/controller.go:282]: Fixed

File: deckhouse-controller/pkg/controller/module-controllers/release/controller.go (hunk 1 above)
Comment (libmonsoon-dev) [reply deckhouse-controller/pkg/controller/module-controllers/release/controller.go:185]: I don't see the point in creating a separate variable when we can just use [named result](https://go.dev/doc/effective_go#named-results)

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @libmonsoon-dev. I've started the workflow for label 'e2e/run/openstack'...

File: deckhouse-controller/pkg/controller/module-controllers/release/controller.go (hunk 1 above)
Comment (ldmonster) [reply deckhouse-controller/pkg/controller/module-controllers/release/controller.go:185]: > I don't see the point in creating a separate variable when we can just use [named result](https://go.dev/doc/effective_go#named-results)

then we need to fix 200 line https://github.com/deckhouse/deckhouse/pull/10105/files#diff-899346a0481b3853b6ef509ea9f5895c648bda8532399ef3938ad02fb22cf008R200

File: deckhouse-controller/pkg/controller/module-controllers/release/controller.go [hunk 7]
```diff
@@ -150,87 +151,82 @@ func NewModuleReleaseController(
...
+func (r *moduleReleaseReconciler) createOrUpdateReconcile(ctx context.Context, mr *v1alpha1.ModuleRelease) (res ctrl.Result, err error) {
 	switch mr.Status.Phase {
 	case "":
 		mr.Status.Phase = v1alpha1.PhasePending
-		mr.Status.TransitionTime = metav1.NewTime(c.dc.GetClock().Now().UTC())
-		if err := c.client.Status().Update(ctx, mr); err != nil {
-			return res, err
+		mr.Status.TransitionTime = metav1.NewTime(r.dc.GetClock().Now().UTC())
+		if err := r.client.Status().Update(ctx, mr); err != nil {
+			return ctrl.Result{}, fmt.Errorf("update status: %w", err)
```
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/module-controllers/release/controller.go:229]: if we use named return parameter, please don't change alias

Comment (libmonsoon-dev) [reply deckhouse-controller/pkg/controller/module-controllers/release/controller.go]: I'm not quite sure what that means in this context. Are we talking about shadowing the err variable?

Comment (ldmonster) [reply deckhouse-controller/pkg/controller/module-controllers/release/controller.go]: you change 
```
return res, err
```
for
```
return ctrl.Result{}, fmt.Errorf("update status: %w", err)
```

Comment (libmonsoon-dev) [reply deckhouse-controller/pkg/controller/module-controllers/release/controller.go]: done

**Review by ldmonster — DISMISSED** [review]

**Review by ldmonster — APPROVED** [review]

**Review by yalosev — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've started the workflow for label 'e2e/run/aws'...

**Pull request merged**